CONSOLE-3005: Provide alias for dynamic plugin's proxy (#967)
openshift/enhancements: enhancements/console/dynamic-plugins.md

In case the plugin needs to communicate with some in-cluster service, it can
declare a service proxy in its `ConsolePlugin` resource using the
`spec.proxy.services` map. The key of the map is the proxy alias
of the service and value is an object in which service `name`, `namespace` and `port`
needs to be specified and to which the request will be proxied.

```yaml
spec:
  proxy:

>     services:

Reviewer: I wonder if we need `services` here now. If we want to add other types of proxies, we want the alias to be unique. I'd suggest:

```yaml
spec:
  proxy:
    my-alias:
      service:
         name: my-name
         namespace: my-namespace
         port: my-port
      caCertificate: ...
      authorize: true
```

Then later on we could do something like

```yaml
spec:
  proxy:
    my-alias:
       url: https://example.com/
       caCertificate: ...
```

Author: So after talking to @stlaz on the API design here we agreed that best would be to have the `spec.proxy` structured as an array, which would contain `alias`, `authorize`, `caCertificate` and  enum `type` field. 
Based on the `type` a proxy config struct would be used.

eg.
```go
type ConsolePluginSpec struct {
	DisplayName string `json:"displayName"`
	Service ConsolePluginService `json:"service"`
	Proxy []ConsolePluginProxy `json:"proxy"`
}

type ConsolePluginProxy struct {
	Type ProxyType `json:"type"`
	Service ProxyServiceConfig `json:"service"`
	URL ProxyTypeURLConfig `json:"url"`
	CACertificate string `json:"caCertificate"`
	Authorize bool `json:"authorize"`
}

type ProxyType string

const (
	ProxyTypeService ProxyType = "Service"
	ProxyTypeURL ProxyType = "URL"
)

type ProxyTypeServiceConfig struct {
	Name string `json:"name"`
	Namespace string `json:"namespace"`
	Port int32 `json:"port"`
}

type ProxyTypeURLConfig struct {
	URL string `json:"url"`
}
```

This is the idiomatic way how the API should be structured... check[ indentityProviders API for example
](https://github.com/openshift/api/blob/9b3afd95c93af476bec246b2c60f64235f078198/config/v1/types_oauth.go#L164-L239)

Reviewer: @jhadvig That looks good to me.

>       <proxy-alias>:

Reviewer: Nit: this is really an alias for the service rather than for the proxy itself.

Author: then I guess it should be `service-proxy-alias` ?

        name: <service-name>
        namespace: <service-namespace>
        port: <service-port>
```

Console backend exposes following endpoint in order to proxy the communication
between plugin and the service:

> `/api/proxy/namespace/<plugin-name>/<proxy-alias>/<request-path>?<optional-query-parameters>`

Reviewer: Remove namespace from the URL?

I'd give it another path element so we can add something else under proxy/ if we want later.

```suggestion
`/api/proxy/plugin/<plugin-name>/<proxy-alias>/<request-path>?<optional-query-parameters>`
```

Author: sorry wanted to remove the `namespace`

> An example proxy request path from `acm-console-plugin` plugin with an
> `acm-proxy` alias is:
> `/api/proxy/acm-console-plugin/acm-proxy/releases?limit=10`

Reviewer: ```suggestion
An example proxy request path from the `acm` plugin with a
`search` service is:
`/api/proxy/plugin/acm/search/pods?namespace=openshift-apiserver`
```

apiVersion: console.openshift.io/v1alpha1
kind: ConsolePlugin
metadata:

>   name: acm-console-plugin

Reviewer: I'd like to discourage plugin authors from putting `-plugin` at the end of all the plugin names :) Let's try to remove that from the examples.

```suggestion
  name: acm
```

spec:

>   displayName: 'ACM console plugin'

Reviewer: ```suggestion
  displayName: 'Advanced Cluster Management'
```

  service:
    name: acm-console-plugin

>     namespace: acm-console-plugin-namespace

Reviewer: ```suggestion
    namespace: open-cluster-management
```

    port: 8443
    basePath: '/'
  proxy:
    services:

>       acm-proxy:
>         name: helm-charts
>         namespace: helm
>         port: 8443
>         caCertificate: '-----BEGIN CERTIFICATE-----\nMIID....'
>         authorize: true

Reviewer: Let's give this a more realistic example.

```suggestion
      search:
        name: search
        namespace: open-cluster-management
        port: 8443
        caCertificate: '-----BEGIN CERTIFICATE-----\nMIID....'
        authorize: true
```

None

> ### API Extensions

Reviewer: Let's override the job instead of adding these sections per the guidance here:

https://github.com/openshift/enhancements/#template-updates
